master: add document for partial index

YangKeao opened the pull request:
### First-time contributors' checklist <!--Remove this section if you're not a first-time contributor.-->

- [x] I've signed the [**Contributor License Agreement**](https://cla.pingcap.net/pingcap/docs), which is required for the repository owners to accept my contribution.

### What is changed, added or deleted? (Required)

1. Add the document for partial index.

### Which TiDB version(s) do your changes apply to? (Required)

<!-- Fill in "x" in [] to tick the checkbox below.-->

**Tips for choosing the affected version(s):**

By default, **CHOOSE MASTER ONLY** so your changes will be applied to the next TiDB major or minor releases. If your PR involves a product feature behavior change or a compatibility change, **CHOOSE THE AFFECTED RELEASE BRANCH(ES) AND MASTER**.

For details, see [tips for choosing the affected versions](https://github.com/pingcap/docs/blob/master/CONTRIBUTING.md#guideline-for-choosing-the-affected-versions).

- [x] master (the latest development version)
- [ ] v9.0 (TiDB 9.0 versions)
- [x] v8.5 (TiDB 8.5 versions)
- [ ] v8.1 (TiDB 8.1 versions)
- [ ] v7.5 (TiDB 7.5 versions)
- [ ] v7.1 (TiDB 7.1 versions)
- [ ] v6.5 (TiDB 6.5 versions)
- [ ] v6.1 (TiDB 6.1 versions)
- [ ] v5.4 (TiDB 5.4 versions)

### What is the related PR or file link(s)?

<!--Reference link(s) will help reviewers review your PR quickly.-->

- This PR is translated from: https://github.com/pingcap/docs-cn/pull/20999
- Other reference link(s):

### Do your changes match any of the following descriptions?

- [ ] Delete files
- [ ] Change aliases
- [ ] Need modification after applied to another branch <!-- If yes, please comment "/label version-specific-changes-required" below to trigger the bot to add the label. -->
- [ ] Might cause conflicts after applied to another branch

YangKeao committed:
add docs for partial index

Signed-off-by: Yang Keao <yangkeao@chunibyo.icu>

winoros commented on sql-statements/sql-statement-create-index.md:
```diff
@@ -366,6 +367,138 @@ See [Index Selection - Use multi-valued indexes](/choose-index.md#use-multi-valu
+
+### Limitations
+
+- The `WHERE` clause in partial indexes supports basic comparison operators (`=`, `!=`, `<`, `<=`, `>`, `>=`) and `IN` predicates with constant values
```
IS (NOT) NULL is not mentioned?

winoros commented on sql-statements/sql-statement-create-index.md:
```diff
@@ -366,6 +367,138 @@ See [Index Selection - Use multi-valued indexes](/choose-index.md#use-multi-valu
+Partial indexes offer several advantages:
+
+1. **Reduced storage**: Only rows matching the predicate are indexed, saving storage space
+2. **Faster DML**: It'll be faster to maintain the index of a subset of data during INSERT, UPDATE, and DELETE operations
```
Something like `3. Ability to maintain a unique constraint on a subset of the table rows`?

YangKeao replied:
It has been mentioned in the `When to use partial indexes` section. It's not part of the `Performance benefits`.

Oreoxmt commented on sql-statements/sql-statement-create-index.md:
```diff
@@ -366,6 +367,138 @@ See [Index Selection - Use multi-valued indexes](/choose-index.md#use-multi-valu
 - If a table uses multi-valued indexes, you cannot back up, replicate, or import the table using BR, TiCDC, or TiDB Lightning to a TiDB cluster earlier than v6.6.0.
 - For a query with complex conditions, TiDB might not be able to select multi-valued indexes. For information on the condition patterns supported by multi-valued indexes, refer to [Use multi-valued indexes](/choose-index.md#use-multi-valued-indexes).
 
+## Partial indexes
```
```suggestion
## Partial indexes <span class="version-mark">New in v8.5.7</span>
```

Oreoxmt commented on sql-statements/sql-statement-create-index.md:
```diff
@@ -366,6 +367,138 @@ See [Index Selection - Use multi-valued indexes](/choose-index.md#use-multi-valu
 - For a query with complex conditions, TiDB might not be able to select multi-valued indexes. For information on the condition patterns supported by multi-valued indexes, refer to [Use multi-valued indexes](/choose-index.md#use-multi-valued-indexes).
 
+## Partial indexes
+
```
```suggestion

Starting from v8.5.7, TiDB supports creating partial indexes.

```

Oreoxmt commented on sql-statements/sql-statement-create-index.md:
```diff
@@ -366,6 +367,138 @@ See [Index Selection - Use multi-valued indexes](/choose-index.md#use-multi-valu
 
+## Partial indexes
+
+A partial index is an index built on a subset of rows in a table, defined by a conditional expression (called the predicate of the partial index). The index contains entries only for those rows that satisfy the predicate.
```
```suggestion
A partial index is an index built on a subset of rows in a table. A conditional expression, called a predicate, defines the subset. The index contains entries only for rows that satisfy the predicate.
```

Oreoxmt commented on sql-statements/sql-statement-create-index.md:
````diff
@@ -366,6 +367,138 @@ See [Index Selection - Use multi-valued indexes](/choose-index.md#use-multi-valu
+ALTER TABLE t1 ADD INDEX idx2 (c1, c2) WHERE c3 = 'abc';
+```
+
+Or specify the partial index when creating the table:
````
```suggestion
You can also specify a partial index when you create the table:
```

Oreoxmt commented on sql-statements/sql-statement-create-index.md:
```diff
@@ -366,6 +367,138 @@ See [Index Selection - Use multi-valued indexes](/choose-index.md#use-multi-valu
+CREATE TABLE users (
+    id INT PRIMARY KEY AUTO_INCREMENT,
+    name VARCHAR(100),
+    status varchar(20),
```
```suggestion
    status VARCHAR(20),
```

Oreoxmt commented on sql-statements/sql-statement-create-index.md:
````diff
@@ -366,6 +367,138 @@ See [Index Selection - Use multi-valued indexes](/choose-index.md#use-multi-valu
+
+Then the following queries can use the partial index:
+
+```
````
```suggestion
```sql
```

Oreoxmt commented on sql-statements/sql-statement-create-index.md:
````diff
@@ -366,6 +367,138 @@ See [Index Selection - Use multi-valued indexes](/choose-index.md#use-multi-valu
+Then the following queries can use the partial index:
+
+```
+mysql> explain SELECT * FROM users WHERE status = 'active' AND name = 'John';
````
```suggestion
mysql> EXPLAIN SELECT * FROM users WHERE status = 'active' AND name = 'John';
```

Oreoxmt commented on sql-statements/sql-statement-create-index.md:
```diff
@@ -366,6 +367,138 @@ See [Index Selection - Use multi-valued indexes](/choose-index.md#use-multi-valu
++-------------------------------+---------+-----------+-------------------------------------------+-------------------------------------------------------+
+4 rows in set (0.00 sec)
+
+mysql> explain SELECT * FROM users WHERE status = 'active' ORDER BY name;
```
```suggestion
mysql> EXPLAIN SELECT * FROM users WHERE status = 'active' ORDER BY name;
```

Oreoxmt commented on sql-statements/sql-statement-create-index.md:
```diff
@@ -366,6 +367,138 @@ See [Index Selection - Use multi-valued indexes](/choose-index.md#use-multi-valu
++-------------------------------+----------+-----------+-------------------------------------------+---------------------------------+
+4 rows in set (0.00 sec)
+
+mysql> explain SELECT * FROM users WHERE score > 10000 ORDER BY created_at;
```
```suggestion
mysql> EXPLAIN SELECT * FROM users WHERE score > 10000 ORDER BY created_at;
```

Oreoxmt commented on sql-statements/sql-statement-create-index.md:
```diff
@@ -366,6 +367,138 @@ See [Index Selection - Use multi-valued indexes](/choose-index.md#use-multi-valu
++-------------------------------+----------+-----------+-----------------------------------------------------+--------------------------------+
+4 rows in set (0.00 sec)
+
+mysql> explain SELECT * FROM users WHERE status = 'pending';
```
```suggestion
mysql> EXPLAIN SELECT * FROM users WHERE status = 'pending';
```

Oreoxmt commented on sql-statements/sql-statement-create-index.md:
````diff
@@ -366,6 +367,138 @@ See [Index Selection - Use multi-valued indexes](/choose-index.md#use-multi-valu
+If the predicates in query don't meet the index definition, the index will not be chosen even with hint:
+
+```
+mysql> explain SELECT * FROM users use index(idx_high_score_users) WHERE score > 100 ORDER BY created_at;
````
```suggestion
mysql> EXPLAIN SELECT * FROM users USE INDEX(idx_high_score_users) WHERE score > 100 ORDER BY created_at;
```

Oreoxmt commented on sql-statements/sql-statement-create-index.md:
````diff
@@ -366,6 +367,138 @@ See [Index Selection - Use multi-valued indexes](/choose-index.md#use-multi-valu
+3 rows in set (0.00 sec)
+```
+
+If the predicates in query don't meet the index definition, the index will not be chosen even with hint:
````
```suggestion
If the query predicates do not match the partial index definition, TiDB does not select the index, even with a hint:
```

Oreoxmt commented on sql-statements/sql-statement-create-index.md:
```diff
@@ -366,6 +367,138 @@ See [Index Selection - Use multi-valued indexes](/choose-index.md#use-multi-valu
+Partial indexes are particularly useful in the following scenarios:
+
+- **Selective filtering**: When you frequently query a small subset of rows based on specific conditions
+- **Conditional uniqueness**: When you need unique constraints that only apply under certain conditions
```
```suggestion
- **Selective filtering**: use a partial index when you frequently query a small subset of rows based on specific conditions.
- **Conditional uniqueness**: use a partial index when you need unique constraints that apply only under certain conditions.
```

Oreoxmt commented on sql-statements/sql-statement-create-index.md:
```diff
@@ -366,6 +367,138 @@ See [Index Selection - Use multi-valued indexes](/choose-index.md#use-multi-valu
+- The `WHERE` clause in partial indexes supports basic comparison operators (`=`, `!=`, `<`, `<=`, `>`, `>=`) and `IN` predicates with constant values
+- The types of the column and constant value should be the same
+- The predicate can only reference columns from the same table
+- Partial indexes cannot be used on expression indexes
```
```suggestion
- The `WHERE` clause in partial indexes supports basic comparison operators (`=`, `!=`, `<`, `<=`, `>`, `>=`) and `IN` predicates with constant values.
- The column and constant value must have the same type.
- The predicate can only reference columns from the same table.
- Partial indexes cannot be used on expression indexes.
```

Oreoxmt commented on sql-statements/sql-statement-create-index.md:
```diff
@@ -366,6 +367,138 @@ See [Index Selection - Use multi-valued indexes](/choose-index.md#use-multi-valu
+
+### Performance benefits
+
+Partial indexes offer several advantages:
```
```suggestion
Partial indexes have the following performance benefits:
```

Oreoxmt commented on sql-statements/sql-statement-create-index.md:
```diff
@@ -366,6 +367,138 @@ See [Index Selection - Use multi-valued indexes](/choose-index.md#use-multi-valu
+Partial indexes offer several advantages:
+
+1. **Reduced storage**: Only rows matching the predicate are indexed, saving storage space
+2. **Faster DML**: It'll be faster to maintain the index of a subset of data during INSERT, UPDATE, and DELETE operations
```
```suggestion
- **Reduced storage**: only the rows that match the predicate are indexed, which saves storage space.
- **Faster DML**: maintaining an index for a subset of rows is faster during `INSERT`, `UPDATE`, and `DELETE` operations.
```

Oreoxmt reviewed (approved):
Rest LGTM

YangKeao committed:
Apply suggestions from code review

Co-authored-by: Oreoxmt <aolinz@outlook.com>

YangKeao committed:
add IS NOT NULL and IS NULL in the limitation

Signed-off-by: Yang Keao <yangkeao@chunibyo.icu>

YangKeao committed:
add language to avoid CI error

Signed-off-by: Yang Keao <yangkeao@chunibyo.icu>

qiancai commented on sql-statements/sql-statement-create-index.md:
````diff
@@ -366,6 +367,140 @@ See [Index Selection - Use multi-valued indexes](/choose-index.md#use-multi-valu
+3 rows in set (0.00 sec)
+```
+
+If the query predicates do not match the partial index definition, TiDB does not select the index, even with a hint:
````
```suggestion
If a query predicate does not match the partial index definition, TiDB does not select the partial index, even with a hint. For example, the following statement cannot use the partial index `idx_high_score_users`, because the query predicate `score > 100` does not match the partial index definition `score > 1000`:
```

qiancai commented on sql-statements/sql-statement-create-index.md:
```diff
@@ -366,6 +367,140 @@ See [Index Selection - Use multi-valued indexes](/choose-index.md#use-multi-valu
+
+### Limitations
+
+- The `WHERE` clause in partial indexes supports basic comparison operators (`=`, `!=`, `<`, `<=`, `>`, `>=`), `IS NULL`, `IS NOT NULL` and `IN` predicates with constant values.
```
```suggestion
- The `WHERE` clause in partial indexes supports basic comparison operators (`=`, `!=`, `<`, `<=`, `>`, `>=`), `IS NULL`, `IS NOT NULL`, and `IN` predicates with constant values.
```

qiancai commented on sql-statements/sql-statement-create-index.md:
```diff
@@ -366,6 +367,140 @@ See [Index Selection - Use multi-valued indexes](/choose-index.md#use-multi-valu
+### Limitations
+
+- The `WHERE` clause in partial indexes supports basic comparison operators (`=`, `!=`, `<`, `<=`, `>`, `>=`), `IS NULL`, `IS NOT NULL` and `IN` predicates with constant values.
+- The column and constant value must have the same type.
```
```suggestion
- The column and constant values must be of the same data type.
```

qiancai commented on sql-statements/sql-statement-create-index.md:
```diff
@@ -366,6 +367,140 @@ See [Index Selection - Use multi-valued indexes](/choose-index.md#use-multi-valu
+- The `WHERE` clause in partial indexes supports basic comparison operators (`=`, `!=`, `<`, `<=`, `>`, `>=`), `IS NULL`, `IS NOT NULL` and `IN` predicates with constant values.
+- The column and constant value must have the same type.
+- The predicate can only reference columns from the same table.
+- Partial indexes cannot be used on expression indexes.
```
question: used on or created on?

YangKeao replied:
Created on. I'll modify it.

YangKeao committed:
Apply suggestions from code review

Co-authored-by: Grace Cai <qqzczy@126.com>

YangKeao committed:
modify examples and limitations

Signed-off-by: Yang Keao <yangkeao@chunibyo.icu>

qiancai commented on sql-statements/sql-statement-create-index.md:
```diff
@@ -366,6 +367,140 @@ See [Index Selection - Use multi-valued indexes](/choose-index.md#use-multi-valu
 
+## Partial indexes <span class="version-mark">New in v8.5.7</span>
+
+Starting from v8.5.7, TiDB supports creating partial indexes.
```
```suggestion
## Partial indexes <span class="version-mark">New in v8.5.7 and v9.0.0</span>
```

qiancai committed:
Update sql-statements/sql-statement-create-index.md

github-actions[bot] committed:
Auto-sync: Update English docs from Chinese PR

Synced from: https://github.com/pingcap/docs-cn/pull/20999/files/c9e411192407215955b3d87e4da49509be70abbb..e4602c870ceb75aea0762a3b8c97d68abb75cb29
Target PR: https://github.com/pingcap/docs/pull/21903
AI Provider: azure

Co-authored-by: github-actions[bot] <github-actions[bot]@users.noreply.github.com>

qiancai committed:
minor wording updates

qiancai committed:
fix a typo

qiancai commented on sql-statements/sql-statement-create-index.md:
```diff
@@ -366,6 +367,136 @@ See [Index Selection - Use multi-valued indexes](/choose-index.md#use-multi-valu
 
+## Partial indexes <span class="version-mark">New in v8.5.7 and v9.0.0</span>
+
+A partial index is an index built on a subset of rows in a table. When creating a partial index, you can specify a conditional expression, also known as a predicate, to define the subset. The index contains entries only for rows that satisfy the predicate.
```
```suggestion
A partial index is an index built on a subset of rows in a table. When creating a partial index, you can specify a conditional expression, also known as a predicate, to define that subset of rows. The index contains entries only for the rows that satisfy the predicate.
```

qiancai committed:
minor wording updates
